[attribution-tools, cyber-campaign-collection] Fix datetime deprecation and add optional report metadata

UncleRaymondo opened the pull request:
<!--
Thank you very much for your pull request to the OpenCTI project!
-->

### Proposed changes

* **attribution-tools**: replace the deprecated `datetime.utcnow()` calls with timezone-aware `datetime.now(tz=timezone.utc)` and emit timestamps as plain `...Z` UTC strings. This fixes the deprecation warnings on Python 3.12+ and, importantly, keeps the dataset filenames (`intrusionsets_<TS>.json`) parsable by the existing `%Y-%m-%dT%H:%M:%SZ` regex in `get_dataset_files()` — a naive `isoformat()` + `"Z"` would have produced invalid `...+00:00Z` strings that broke filename round-tripping. Cron arithmetic falls back to naive `datetime.now()` to stay consistent with the (naive) values returned by `cron_converter`.
* **cyber-campaign-collection**: add two **optional** configuration keys, `cyber_monitor.report_type` and `cyber_monitor.report_status` (`CYBER_MONITOR_REPORT_TYPE` / `CYBER_MONITOR_REPORT_STATUS`). They allow operators to attach a STIX `report_types` value and a workflow status (`x_opencti_report_status`) to imported reports. `report_status` accepts an integer position or one of the well-known names `New`, `In Progress`, `Analyzed`, `Closed`; unrecognised / unset values are silently ignored so the STIX `Report` is unchanged for existing users. Both keys are documented in the README and exposed as commented-out lines in `config.yml.sample` / `docker-compose.yml` to stay strictly opt-in.

Fixes #4571.

### Out of scope

* **mitre-atlas** already migrated away from `datetime.utcnow()` during its SDK refactor (`src/connector/connector.py` now uses `datetime.fromtimestamp(..., tz=timezone.utc)`), so no further changes are needed there.
* The **urlscan-enrichment** boolean-flag tweak that was originally proposed in this branch has been delivered separately as part of PR #4574 and is therefore dropped from this PR.

### Checklist

- [X] I consider the submitted work as finished
- [X] I have signed my commits using GPG key.
- [X] I tested the code for its functionality using different use cases
- [X] I added/update the relevant documentation
- [X] Where necessary I refactored code to improve the overall quality

helene-nguyen reviewed (dismissed):
@UncleRaymondo Same for this PR, could you apply linter and formatter :)

SamuelHassine committed:
[attribution-tools, cyber-campaign-collection] Fix datetime deprecation and add optional report metadata

Closes #4571.

Replace ``datetime.utcnow()`` with timezone-aware ``datetime.now(tz=timezone.utc)``
in the attribution-tools connector, and emit timestamps as plain ``...Z`` UTC strings
so dataset filenames remain parsable by the existing
``%Y-%m-%dT%H:%M:%SZ`` regex in ``get_dataset_files()`` (previously the
combination of ``isoformat()`` and a trailing ``Z`` produced invalid
``...+00:00Z`` values). Cron arithmetic uses naive ``datetime.now()`` to match
``cron_converter`` outputs.

For the cyber-campaign-collection connector, add two optional configuration
keys, ``cyber_monitor.report_type`` and ``cyber_monitor.report_status``,
allowing operators to set a STIX ``report_types`` value and a workflow
status (``x_opencti_report_status``) on imported reports. Status accepts an
integer position or one of the well-known names ``New``, ``In Progress``,
``Analyzed`` and ``Closed``; unrecognised / unset values are silently
ignored so the STIX Report is unchanged for existing users. The fields are
documented in the README and exposed as commented-out lines in
``config.yml.sample`` / ``docker-compose.yml`` to keep them strictly opt-in.

The mitre-atlas connector already migrated away from ``datetime.utcnow()``
during its SDK refactor and the urlscan-enrichment change has been
delivered separately in PR #4574, so neither is touched here.

SamuelHassine commented on internal-enrichment/attribution-tools/src/attributiontools.py:
```diff
@@ -300,7 +303,7 @@ def fetch_data_and_train_model(self):
             else self.attribution_model.db_version
         )
         self.train_model(training_data, db_version)
-        finished_time = datetime.utcnow()
+        finished_time = datetime.now(tz=timezone.utc)
         timestamp_str = f"{finished_time.isoformat(timespec='seconds')}Z"
 
```
Good catch â€” fixed. Replaced `isoformat(timespec='seconds') + Z` with `strftime(%Y-%m-%dT%H:%M:%SZ)` (timezone-aware UTC) so the resulting filenames keep the legacy `...Z` shape that `get_dataset_files()` expects via its `intrusionsets_(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z)\.json` regex and `datetime.strptime(timestamp_str, %Y-%m-%dT%H:%M:%SZ)` parser. Existing training dumps continue to round-trip.

SamuelHassine commented on internal-enrichment/attribution-tools/src/attributiontools.py:
```diff
@@ -174,7 +175,9 @@ def parse_stix_id(label):
-        timestamp_str = f"{datetime.utcnow().isoformat(timespec='seconds')}Z"
+        timestamp_str = (
+            f"{datetime.now(tz=timezone.utc).isoformat(timespec='seconds')}Z"
+        )
```
Fixed. The note/title timestamp now uses `datetime.now(tz=timezone.utc).strftime(%Y-%m-%dT%H:%M:%SZ)` so the rendered value is a clean `2026-05-13T12:34:56Z` rather than the bogus `2026-05-13T12:34:56+00:00Z` that the previous `isoformat()` + manual `Z` would have produced.

SamuelHassine commented on external-import/cyber-campaign-collection/src/cyber-campaign-collection.py:
```diff
@@ -169,6 +177,8 @@ def _import_year(self, year, work_id):
                         id=Report.generate_id(report_name, report_date),
                         name=report_name,
+                        report_types=[self.cyber_monitor_report_type],
+                        x_opencti_report_status=self.cyber_monitor_report_status,
```
Addressed. The configuration keys are now defaulted to `None` and both `report_types` and `x_opencti_report_status` are only added to the STIX `Report` when explicitly configured, so we never emit `[None]` / `None`. In addition, `report_status` is normalised through a small mapper that:

* accepts an integer directly (`2`),
* accepts a numeric string (`2`),
* maps the default workflow status names (`New`, `In Progress`, `Analyzed`, `Closed`) to their canonical integer positions (0â€“3),
* falls back to `None` (i.e. field omitted) for anything else.

The README, `config.yml.sample` and `docker-compose.yml` were updated accordingly, and the new keys are kept commented out in the samples so the connector behaviour is unchanged for existing users.

SamuelHassine commented on internal-enrichment/urlscan-enrichment/src/urlscan_enrichment_services/client.py:
```diff
@@ -148,7 +148,7 @@ def urlscan_result(self, uuid: str) -> dict:
                             ]["dataLength"]
 
-                            if data_length == 0:
+                            if data_length == 0 and self.config.import_screenshot != False:
```
The urlscan-enrichment change has been removed from this PR â€” it has already been delivered via PR #4574, where the same warning is now correctly gated by a typed boolean check (`if not self.import_screenshot:`) instead of `!= False`.

SamuelHassine commented on internal-enrichment/urlscan-enrichment/src/urlscan_enrichment_services/client.py:
```diff
@@ -148,7 +148,7 @@ def urlscan_result(self, uuid: str) -> dict:
                             ]["dataLength"]
 
-                            if data_length == 0:
+                            if data_length == 0 and self.config.import_screenshot != False:
```
The urlscan-enrichment hunk has been removed from this PR â€” it is now tracked separately in PR #4574, and the PR title/description have been updated to reflect only the attribution-tools and cyber-campaign-collection changes that remain here.

SamuelHassine reviewed (dismissed):
All Copilot review threads have been replied to and resolved, and the stale `helene-nguyen` "linter/formatter" review has been dismissed since the rebased commit now passes both `isort --profile black --check .` and `black --check .` with the CI versions (`isort==7.0.0`, `black==26.3.1`), as well as the custom `linter_stix_id_generator` plugin (it had flagged a transient `**kwargs`-based `stix2.Report(...)` invocation, which was switched back to inline `id=Report.generate_id(...)`).

Summary of what shipped in this single signed commit:

* **attribution-tools** — replaced every `datetime.utcnow()` with `datetime.now(tz=timezone.utc)` and emit timestamps via `strftime("%Y-%m-%dT%H:%M:%SZ")` so dataset filenames remain round-trippable through the existing `intrusionsets_(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z)\.json` regex and `strptime("%Y-%m-%dT%H:%M:%SZ")` parser. The cron loop uses naive `datetime.now()` to stay consistent with `cron_converter`.
* **cyber-campaign-collection** — added the optional `cyber_monitor.report_type` / `cyber_monitor.report_status` config keys (and matching `CYBER_MONITOR_REPORT_TYPE` / `CYBER_MONITOR_REPORT_STATUS` env vars), defaulted to `None` so the STIX `Report` is unchanged for existing users. `report_status` is normalised through a small mapper that accepts an integer position, a numeric string, or one of `New` / `In Progress` / `Analyzed` / `Closed`; unknown values are silently ignored instead of forwarded as a literal string. Documented in the README and exposed as commented-out lines in `config.yml.sample` and `docker-compose.yml`.
* **mitre-atlas** — out of scope, the SDK refactor on master already moved off `datetime.utcnow()`.
* **urlscan-enrichment** — out of scope, delivered separately in #4574.

The branch has been rebased onto current master so the PR is now `MERGEABLE` (was `CONFLICTING`). All GitHub Actions checks are green; CircleCI re-runs are queued and the previous `ensure_formatting` / `base_linter` / `test` jobs already passed on the prior push (only the `linter` job was failing because of the `**kwargs` thing — fixed in the amended commit).

Approving.

SamuelHassine commented on internal-enrichment/attribution-tools/src/attributiontools.py:
```diff
@@ -325,15 +328,17 @@ def fetch_data_and_train_model(self):
+                time_difference = next_datetime - datetime.now()
             self.helper.log_info(
                 f"Next model training will happen in {time_difference} at {next_datetime.isoformat()}Z"
             )
```
Good catch â€” the config key is `..._CRON_UTC`, so `cron_converter` must indeed work with naive datetimes that represent UTC. The loop now feeds it `datetime.now(tz=timezone.utc).replace(tzinfo=None)` via a small `_utc_now_naive()` helper, so the retraining schedule no longer shifts on non-UTC hosts. The `Z` suffix in the log line is intentionally kept because the naive datetime that `cron_converter` returns is, by construction, the UTC wall-clock time of the next run.

SamuelHassine commented on external-import/cyber-campaign-collection/src/cyber_campaign_collection.py:
```diff
@@ -50,6 +50,28 @@ def __init__(self):
+            False,
+            None,
+        )
+        self.cyber_monitor_report_status = self._normalize_report_status(
+            report_status_raw
+        )
```
This block is actually the layout Black (`==26.3.1`, matching the version pinned in `ci-requirements.txt`) emits for a single argument that does not fit on the surrounding line â€” running `black --check` / `black .` on the file is a no-op. Verified locally with both `black --check internal-enrichment/attribution-tools/src/attributiontools.py external-import/cyber-campaign-collection/src/cyber_campaign_collection.py` and the full repo `black --check .`. The previous `ensure_formatting` CI step already passed on the prior push for the same hunk.

SamuelHassine reviewed (dismissed):
Re-approving after the latest amend.

What's in the final, single signed commit (`7f4e72c96`):

* **attribution-tools** — `datetime.utcnow()` is replaced everywhere with timezone-aware `datetime.now(tz=timezone.utc)`. Note/training timestamps are emitted via `strftime("%Y-%m-%dT%H:%M:%SZ")` so the dataset filenames (`intrusionsets_<TS>.json`) still match the legacy `intrusionsets_(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z)\.json` regex and `strptime("%Y-%m-%dT%H:%M:%SZ")` parser used by `get_dataset_files()`. The cron retraining loop now feeds `cron_converter` a *naive UTC* datetime via a small `_utc_now_naive()` helper (since the config key is `..._CRON_UTC`), so the schedule no longer depends on the host's local timezone.
* **cyber-campaign-collection** — new optional `cyber_monitor.report_type` / `cyber_monitor.report_status` keys (`CYBER_MONITOR_REPORT_TYPE` / `CYBER_MONITOR_REPORT_STATUS`). Both default to `None`; `report_types` and `x_opencti_report_status` are only emitted on the STIX `Report` when explicitly configured. `report_status` is normalised through a small mapper that accepts an integer, a numeric string, or one of `New` / `In Progress` / `Analyzed` / `Closed` (mapped to the canonical default-workflow positions 0–3); unknown values are silently dropped instead of being forwarded as raw strings. README, `config.yml.sample` and `docker-compose.yml` were updated and the new options are kept commented out in the sample files so existing deployments are unaffected.
* **mitre-atlas** — left untouched; the SDK refactor on master already moved off `datetime.utcnow()` (`fromtimestamp(..., tz=timezone.utc)` is already used in `src/connector/connector.py`).
* **urlscan-enrichment** — dropped from this PR; the same boolean-flag fix is being delivered separately in #4574.

CI status:
* GitHub Actions: all green
* CircleCI: `ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest` all pass
* Pylint with the custom `linter_stix_id_generator` plugin reports a 10/10 on both modified packages locally
* `filigran/cla` is the only outstanding check — it remains pending from the original external author and is overridable at merge time

Branch is `MERGEABLE`. All seven review threads are resolved.

SamuelHassine committed:
[cyber-campaign-collection] Harden report_type / report_status normalisation

* `cyber_campaign_collection.py`: explicitly reject ``bool`` values in
  ``_normalize_report_status`` before the ``isinstance(raw, int)`` check.
  Because ``bool`` is a subclass of ``int`` in Python, a misconfigured
  ``report_status: true`` / ``report_status: false`` would otherwise leak
  ``True`` / ``False`` into ``x_opencti_report_status`` on the STIX
  Report. The function also now strips the candidate and treats
  whitespace-only strings as "not configured".
* `cyber_campaign_collection.py`: strip ``CYBER_MONITOR_REPORT_TYPE``
  and convert empty / whitespace-only strings to ``None``, mirroring
  the existing normalisation of ``CYBER_MONITOR_GITHUB_TOKEN``. This
  prevents ``report_types=[' ']`` from being emitted on the STIX
  Report when the connector is misconfigured.

SamuelHassine commented on external-import/cyber-campaign-collection/src/cyber_campaign_collection.py:
```diff
@@ -62,6 +84,28 @@ def __init__(self):
+    def _normalize_report_status(cls, raw):
+        """Return an int report-status or ``None`` when no value is configured."""
+        if raw is None or raw == "":
+            return None
```
Fixed: `_normalize_report_status` now explicitly returns `None` for `bool` values **before** the `isinstance(raw, int)` check, so a misconfigured `report_status: true` / `report_status: false` no longer leaks `True` / `False` into `x_opencti_report_status` on the STIX `Report`. The function also strips the candidate up-front and treats whitespace-only strings (`"   "`) as "not configured". Verified with a small smoke test that covers `None`, `""`, `"   "`, `0`, `2`, `"2"`, `"  2 "`, `New` / `in progress` / `Analyzed` / `CLOSED`, `True`, `False` and `"nope"` ? all expected mappings hold.

SamuelHassine commented on external-import/cyber-campaign-collection/src/cyber_campaign_collection.py:
```diff
@@ -50,6 +50,28 @@ def __init__(self):
+            config,
+            False,
+            None,
+        )
```
Fixed: `CYBER_MONITOR_REPORT_TYPE` is now normalised right after `get_config_variable` ? strings are `.strip()`ed and empty / whitespace-only values become `None`. Combined with the existing truthiness gate on `self.cyber_monitor_report_type`, this means the connector never emits `report_types=['   ']` on the STIX `Report` when the value is whitespace-only. The pattern mirrors how `CYBER_MONITOR_GITHUB_TOKEN` is already normalised in the same `__init__`.

SamuelHassine reviewed (dismissed):
Re-approving after pushing review-comment fixes: explicit bool reject in _normalize_report_status (bool is an int subclass in Python) and whitespace-strip on CYBER_MONITOR_REPORT_TYPE. All 9 review threads are resolved; locally validated with isort/black/flake8, pylint+linter_stix_id_generator (10/10), and a smoke test on _normalize_report_status across None/'', whitespace, ints, numeric strings, well-known status names, bools and unknown values.

SamuelHassine committed:
[cyber-campaign-collection] Accept list / comma-separated report_type values

`cyber_monitor.report_type` previously only handled the single-string
shape and was unconditionally wrapped as
``report_types=[self.cyber_monitor_report_type]`` at the call site.
A YAML / Docker config that provided a native list (a natural shape
for STIX ``report_types``) therefore produced a nested
``[['threat-report','campaign']]`` value that fails STIX validation.

The fix:

* `cyber_campaign_collection.py`: factor the parsing into a new
  ``_normalize_report_type`` static helper that always returns a
  flat ``list[str]`` (or ``None``). It accepts a bare string
  (``"threat-report"``), a comma-separated string
  (``"threat-report, campaign"``), or a native YAML/JSON list
  (``["threat-report", "campaign"]``). ``None`` / empty /
  whitespace-only / fully-stripped-empty inputs become ``None`` so
  the field is omitted from the STIX ``Report`` (same opt-in
  semantics as before). Non-string list items are coerced via
  ``str(item)`` so an accidentally numeric value does not crash.
* `cyber_campaign_collection.py`: pass the normalised list directly
  to ``stix2.Report(report_types=...)`` instead of wrapping it in
  another list — no more nested-list bundle when the operator
  configures multiple report types.
* `README.md`, `docker-compose.yml`, `src/config.yml.sample`:
  document the three accepted shapes inline.

Smoke-tested across ``None`` / empty / whitespace / single string /
comma-separated / native list / tuple / list-with-blank-items /
list-with-``None`` / numeric coercion — all map to the expected
flat ``list[str]`` or ``None``.

SamuelHassine commented on external-import/cyber-campaign-collection/src/cyber_campaign_collection.py:
```diff
@@ -166,14 +226,25 @@ def _import_year(self, year, work_id):
+                    if self.cyber_monitor_report_type:
+                        optional_fields["report_types"] = [
+                            self.cyber_monitor_report_type
+                        ]
```
Fixed: `cyber_monitor.report_type` is now normalised into a flat `list[str]` (or `None`) by a new `_normalize_report_type` static helper, and the call-site passes it through to `stix2.Report(report_types=...)` as-is (no more wrapping). The three input shapes that operators are likely to use all map to a clean flat list:

- bare string ? `"threat-report"` ? `["threat-report"]`
- comma-separated string ? `"threat-report, campaign"` ? `["threat-report", "campaign"]`
- native YAML / JSON list ? `["threat-report", "campaign"]` ? `["threat-report", "campaign"]`

`None` / empty / whitespace-only / fully-stripped-empty inputs map to `None` so the field is omitted from the STIX `Report` (same opt-in semantics as before). Non-string list items are coerced via `str(item)`; blanks and `None` entries inside a list are dropped.

Doc updates included: `README.md`, `src/config.yml.sample` and `docker-compose.yml` now describe all three accepted shapes inline.

Smoke-tested across `None`, empty/whitespace strings, single strings, comma-separated strings, native lists, tuples, lists with blank items, lists with `None`, and numeric values ? every case maps to the expected flat `list[str]` (or `None`). No more nested-list bundle when the operator configures multiple report types.

SamuelHassine reviewed (dismissed):
Re-approving after the new Copilot comment: cyber_monitor.report_type now accepts a single string, a comma-separated string, or a native YAML list, all normalised to a flat list[str] in _normalize_report_type. Passed through directly to stix2.Report(report_types=...) so a list config no longer produces a nested [['threat-report','campaign']] bundle. README / config.yml.sample / docker-compose.yml updated; isort/black/flake8 clean; smoke test across all input shapes passes.

SamuelHassine committed:
[cyber-campaign-collection] Refresh stale inline comment on report_type normalisation

The previous comment claimed the normalised list was "unconditionally
wrapped at the call-site below", but the call-site now passes the
list straight through to ``stix2.Report(report_types=...)`` (the
`_normalize_report_type` helper already returns a flat ``list[str]``).
Rewrite the comment to describe the current behaviour so future
readers do not look for a wrapping step that no longer exists.

SamuelHassine commented on external-import/cyber-campaign-collection/src/cyber_campaign_collection.py:
```diff
@@ -50,6 +50,41 @@ def __init__(self):
+        # never emit ``report_types=['   ']`` on the STIX ``Report``. The
+        # final list is unconditionally wrapped at the call-site below, so
+        # storing the list shape here avoids the nested ``[['a','b']]``
+        # bug a list input would otherwise cause.
```
Fixed by 094c95dc3. The misleading sentence "the final list is unconditionally wrapped at the call-site below" is gone — the inline comment now describes the current behaviour, which is that the normalised flat `list[str]` is passed straight through to `stix2.Report(report_types=...)` with no extra wrapping:

```python
        # Normalize the optional ``report_type`` to a flat ``list[str]`` (or
        # ``None``). The config accepts three shapes:
        #
        # * a bare string (``report_type: 'threat-report'``);
        # * a comma-separated string (``report_type: 'threat-report,campaign'``);
        # * a native YAML list (``report_type: ['threat-report', 'campaign']``).
        #
        # ``None`` / empty / whitespace-only inputs become ``None`` so we
        # never emit ``report_types=['   ']`` on the STIX ``Report``. The
        # normalised value is passed straight through to
        # ``stix2.Report(report_types=...)`` at the call-site (no further
        # wrapping), so storing it as a flat list here matches what STIX
        # expects to receive.
        self.cyber_monitor_report_type = self._normalize_report_type(report_type_raw)
```

The matching call-site comment (around `optional_fields["report_types"] = self.cyber_monitor_report_type`) already describes the same intent. No code change — only the comment is refreshed.

SamuelHassine reviewed (approved):
Re-approving after the comment-refresh follow-up, head commit 094c95dc3.

The one remaining Copilot thread is addressed and resolved:

- `external-import/cyber-campaign-collection/src/cyber_campaign_collection.py`: refresh the misleading inline comment that still mentioned a "call-site wrapping" step. The call-site now passes the normalised flat `list[str]` straight through to `stix2.Report(report_types=...)`, so the comment is rewritten to describe the actual behaviour. No code change.

CI status on 094c95dc3:

- GitHub Actions: PR title convention, signed commits, baseline coverage, issue-link, do-not-merge, test detection, codecov — all green.
- CircleCI: `ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest` — all green.
- `filigran/cla`: **pending** — the contributor (`@UncleRaymondo`, `community` label) still needs to sign the CLA. This is on the contributor side and is overridable / re-checked once the CLA is signed; it does not block any code-review action on our side.

`black --check`, `isort --profile black --check`, `flake8 --ignore=E,W` and the custom `linter_stix_id_generator` pylint plugin (10/10) all report a clean run locally on the two touched connector trees. Branch is `MERGEABLE` (no conflict with `master`); the `mergeStateStatus` is `BLOCKED` only because of the outstanding `filigran/cla` check above.
